### PR TITLE
log exception from the op result

### PR DESCRIPTION
**Pull request by cid77**
for java op, remove message from stacktrace; option to disable this behavior

### CODE REVIEW

File: cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/steps/ActionExecutionData.java [hunk 1]
```diff
@@ -193,9 +205,44 @@ public void doAction(@Param(EXECUTION_RUNTIME_SERVICES) ExecutionRuntimeServices
...
+
+        String stacktrace = (String) returnMap.get(EXCEPTION);
+
+        if (REMOVE_MESSAGE_FROM_LOGGED_EX) {
+            StringWriter writer = new StringWriter();
+            final PrintWriter printWriter = new PrintWriter(writer);
+
+            BufferedReader reader = new BufferedReader(new StringReader(stacktrace));
+
+            reader.lines().forEach(line -> {
```
Comment (lucian-cm) [inline cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/steps/ActionExecutionData.java:231]: lambdas should be stateless, should not touch variables from ouside like printWriter, consider having a standard for

File: cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/steps/ActionExecutionData.java [hunk 2]
```diff
@@ -193,9 +205,44 @@ public void doAction(@Param(EXECUTION_RUNTIME_SERVICES) ExecutionRuntimeServices
...
+                int idx = line.indexOf(MARKER);
+                String str = idx == -1 ? line : line.substring(0, idx + MARKER.length());
+
+                printWriter.println(str);
+            });
+
+            printWriter.close();
```
Comment (lucian-cm) [inline cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/steps/ActionExecutionData.java:238]: should not close printWriter in memory as it does nothing.

File: cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/steps/ActionExecutionData.java [hunk 3]
```diff
@@ -245,8 +292,14 @@ private String normalizeJavaGav(String gav) {
     private Map<String, Value> prepareAndRunPythonAction(Collection<String> dependencies, String pythonScript,
                                                          Map<String, Value> callArguments, boolean useJython) {
         if (StringUtils.isNotBlank(pythonScript)) {
-            return scriptExecutor.executeScript(
+            final Map<String, Value> returnedMap = scriptExecutor.executeScript(
                     normalizePythonDependencies(dependencies), pythonScript, callArguments, useJython);
+
+            if (returnedMap.containsKey(EXCEPTION)) {
```
Comment (lucian-cm) [inline cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/steps/ActionExecutionData.java:298]: isNotEmpty check seems better as you can have a null or "" value for a key.

File: cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/steps/ActionExecutionData.java [hunk 4]
```diff
@@ -245,8 +292,14 @@ private String normalizeJavaGav(String gav) {
...
+                logger.warn("Action's exception: " + returnedMap.get(EXCEPTION));
```
Comment (lucian-cm) [inline cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/steps/ActionExecutionData.java:299]: I believe we should use error logging level with a better explanation: _"Python operation encountered an exception: "_

File: cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/steps/ActionExecutionData.java (hunk 2 above)
Comment (cid77) [reply cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/steps/ActionExecutionData.java:238]: there may be unflushed data. It doesn't harm either.

File: cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/steps/ActionExecutionData.java [hunk 5]
```diff
@@ -193,9 +205,44 @@ public void doAction(@Param(EXECUTION_RUNTIME_SERVICES) ExecutionRuntimeServices
...
+
+        if (hasException(returnMap)) {
+            logException(returnMap);
+        }
+
         return handleSensitiveValues(returnMap, currentContext);
     }
 
+    private boolean hasException(Map<String, Serializable> returnMap) {
+        return returnMap.containsKey(RETURN_CODE) && !"0".equals(returnMap.get(RETURN_CODE)) &&
```
[the remark this answers is not in the document]
Comment (cid77) [reply cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/steps/ActionExecutionData.java:217]: what's this?

File: cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/steps/ActionExecutionData.java (hunk 3 above)
Comment (cid77) [reply cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/steps/ActionExecutionData.java:298]: the value is not a String as the isNotEmpty function expects

File: cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/steps/ActionExecutionData.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (lucian-cm) [reply cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/steps/ActionExecutionData.java]: exception should be of type String.

**Review by lucian-cm — APPROVED** [review]

**Pull request merged**